### PR TITLE
ci: add exercise testing ci

### PR DESCRIPTION
**Pull request by vaeng**
That is a first for me, adding workflows.
How can we test that?

### CODE REVIEW

File: .github/workflows/test.yml
Comment (IsaacG) [inline .github/workflows/test.yml:3]: The to do/instruction comments could be dropped

File: .github/workflows/test.yml
Comment (IsaacG) [inline .github/workflows/test.yml:33]: This probably needs filling in

File: bin/run.sh
Comment (IsaacG) [inline bin/run.sh:43]:
```suggestion
    jq -n --arg output "${test_output}" '{version: 3, status: "error", message: $output}' > "${results_file}"
```

File: bin/run.sh
Comment (IsaacG) [inline bin/run.sh:36]:
```suggestion
cd "${solution_dir}" || exit
```

File: bin/run.sh
Comment (IsaacG) [inline bin/run.sh:45]: Can you do this with a single jq and a single input/output file?

```
jq ... input > output
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:23]:
```suggestion
    if [ -d "$concept_exercise_dir" ]; then
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:33]:
```suggestion
    if [ -d "$practice_exercise_dir" ]; then
```

**Review by IsaacG — DISMISSED** [review]

File: .github/workflows/test.yml
Comment (vaeng) [reply .github/workflows/test.yml:33]: There is no predefined sqlite action I could find. I better delete it.

**Comment by glennj** [comment]
> How can we test that?

I let github test it for me: create a PR and add commits until it's working.

File: bin/run.sh
Comment (glennj) [inline bin/run.sh:4]: Pedantically, this isn't running the test runner. The 8th track actually uses it's own test runner, and frankly it's a PITA: PRs from forks can't run tests, only PRs from branches in the 8th repo can.

File: bin/verify-exercises
Comment (glennj) [inline bin/verify-exercises:27]:
```suggestion
        cp "${concept_exercise_dir}"/.meta/exemplar.sql "${concept_exercise_dir}/${slug}".sql
        bin/run.sh "${slug}" "${concept_exercise_dir}" "${concept_exercise_dir}"
```

**Review by glennj — CHANGES_REQUESTED** [review]

File: bin/verify-exercises
Comment (glennj) [reply bin/verify-exercises:27]: I have more thoughts about what a verify script should be able to do. But I'll hold those until after this PR is merged.

File: bin/run.sh
Comment (ErikSchierboom) [reply bin/run.sh:4]: Too be fair, that is only because the 8th test runner is special in that the binary is encrypted and thus CI requires access to a secret (which forks don't have access to). 

You could run the sqlite test runner and things should work fine on forks. I think a couple of tracks do this, including Unison. That said, most tracks don't, and that is usually fine as test runners should have smoke tests.

File: bin/run.sh
Comment (vaeng) [reply bin/run.sh]: This was taken from a bad version of a test runner. I don't need the json files, I only need to see if any exercise fails, because the examples should not.

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:1]: How do you feel about upgrading to `bash`?

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:20]:
```suggestion
if [ $# -ne 3 ]; then
    echo "Usage: $0 exercise-slug path/to/solution/folder/ path/to/output/directory/"
```

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:37]: Do you need a DB file here?

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:38]:
```suggestion
[ -f ./user_output.md ] && rm ./user_output.md
```

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:41]: This is testing the `rm` command, not the `sqlite` command :smile:

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:48]:
```suggestion
failed_results=$(jq '.[] | select(.status=="fail") | {name, message}' ./output.json)
```

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:49]:
```suggestion
rm ./output.json
```

If the file must exist, why should a `rm` failure be ignored? If `rm` fails, something is pretty broken.

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:50]:
```suggestion
if [ -n "${failed_results}" ]; then
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:27]: Should the verify script copy the example/examplar file? That seems like it should be part of the verify logic.

Should something restore the `${slug}.sql` file?

File: bin/verify-exercise-example
Comment (vaeng) [reply bin/verify-exercise-example:1]: No feelings at all :D

File: bin/verify-exercise-example
Comment (vaeng) [reply bin/verify-exercise-example:40]: I noticed and moved it :D

File: bin/verify-exercises
Comment (vaeng) [reply bin/verify-exercises:27]: In my understanding, the container is ditched after the testing, so there is no need to preserve the data.

File: bin/verify-exercise-example
Comment (vaeng) [reply bin/verify-exercise-example]: I don't understand. I do not need a db file. But I need the error output for debugging.

File: bin/verify-exercises
Comment (IsaacG) [reply bin/verify-exercises:27]: Maintainers may want to verify things locally, too!

Comment (vaeng) [reply bin/verify-exercises:27]: That's true. I never used the lcoal setup in the c++ track, because it does not fit my work flow. But here we could preserve it.

**Comment by vaeng** [comment]
Hooray, the test ci runs through. But it should not 🥲 

```text
Run bin/verify-exercises
Checking hello-world exercise...
hello-world: testing...
/home/runner/work/sqlite/sqlite/exercises/practice/hello-world
parse error: Invalid numeric literal at line 1, column 4
hello-world: done
Checking leap exercise...
leap: testing...
/home/runner/work/sqlite/sqlite/exercises/practice/leap
jq: error (at ./output.json:1): Cannot iterate over string ("year not d...)
parse error: Expected value before ',' at line 1, column 41
leap: done
Checking raindrops exercise...
raindrops: testing...
/home/runner/work/sqlite/sqlite/exercises/practice/raindrops
jq: error (at ./output.json:1): Cannot iterate over string ("the sound ...)
parse error: Expected value before ',' at line 1, column 23
raindrops: done
```

File: bin/run.sh
Comment (vaeng) [reply bin/run.sh]: I first thought to pull the test runner from its repository, but that is overkill. I don't care for the results.json, I only need to see if any tests fail, so the CI gets an alarm. 

I redid that part and amended the synopsis.

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:19]:
```suggestion
if (( $# != 3 )); then
```

bash features!

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:40]:
```suggestion
if (( $? != 0 )); then
```

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:45]:
```suggestion
[[ -e ./user_output.md ]] && rm ./user_output.md
```

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:48]:
```suggestion
failed_results=$(jq '.[] | select(.status=="fail") | {name, message}' ./output.json) || exit
```

File: bin/verify-exercise-example
Comment (IsaacG) [inline bin/verify-exercise-example:50]:
```suggestion
if [[ -n "${failed_results}" ]]; then
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:23]:
```suggestion
    if [[ -d "$concept_exercise_dir" ]]; then
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:24]:
```suggestion
        slug="${practice_exercise_dir##*/"
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:35]:
```suggestion
    if [[ -d "$practice_exercise_dir" ]]; then
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:36]:
```suggestion
        slug="${practice_exercise_dir##*/}"
```

File: bin/verify-exercises
Comment (IsaacG) [inline bin/verify-exercises:28]: Do you want to collect/count/check if exercises fail the validation?

You can use a count.

```bash
failures=0
...
bin/verify-exercise-example "${slug}" "${concept_exercise_dir}" "${concept_exercise_dir}" && (( failures++ ))
...

if (( failures )); then
  echo "${failures} tests failed!"
  exit 1
fi
exit 0

Comment (vaeng) [reply bin/verify-exercises:28]: Yes, that might be useful.

**Pull request merged**
